src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,433 @@
+            /// 2. Parameter that has default value, and it has no argument. It should be replaced by the default value.
+            /// 3. Parameter mapping to identifier expression
+            /// Before:
+            /// void Caller(int i, int j, bool[] k)
+            /// {
+            ///     Callee(i, j, k);
+            /// }
+            /// void Callee(int a, int b, params bool[] c)
+            /// {
+            ///     DoSomething(a, b, c);
+            /// }
+            /// After:
+            /// void Caller(int i, int j, bool[] k)
+            /// {
+            ///     DoSomething(i, j, k);
+            /// }
+            /// void Callee(int a, int b, params bool[] c)
+            /// {
+            ///     DoSomething(a, b, c);
+            /// }
+            /// 4. A special case, the parameter is only read once in the callee method body
+            /// Before:
+            /// void Caller(bool x)
+            /// {
+            ///     Callee(Foo(), Bar())
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// After:
+            /// void Caller(bool x)
+            /// {
+            ///     DoSomething(Foo(), Bar());
+            /// }
+            /// void Callee(int a, int b)
+            /// {
+            ///     DoSomething(a, b);
+            /// }
+            /// In this case, parameters 'a' and 'b' should just be replaced by the argument expression.
+            /// Note: this might cause semantics changes. It is by design.
+            /// </summary>
+            public ImmutableDictionary<IParameterSymbol, TExpressionSyntax> ParametersToReplace { get; }
+
+            public MethodParametersInfo(
+                ImmutableArray<(IParameterSymbol parameterSymbol, string name)> parametersWithVariableDeclarationArgument,
+                ImmutableArray<(IParameterSymbol parameterSymbol, TExpressionSyntax expression)> parametersToGenerateFreshVariablesFor,
+                ImmutableDictionary<IParameterSymbol, TExpressionSyntax> parametersToReplace)
+            {
+                ParametersWithVariableDeclarationArgument = parametersWithVariableDeclarationArgument;
+                ParametersToGenerateFreshVariablesFor = parametersToGenerateFreshVariablesFor;
+                ParametersToReplace = parametersToReplace;
+            }
+        }
+
+        private async Task<MethodParametersInfo> GetMethodParametersInfoAsync(
+            Document document,
+            TMethodDeclarationSyntax calleeMethodNode,
+            IInvocationOperation invocationOperation,
+            CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var allArgumentOperations = invocationOperation.Arguments
+                .WhereAsArray(argument => argument.Children.IsSingle());
```

Comment:
explain the .IsSingle part.

Reply:
esp. as this seems like you could just be throwing out argumetns... not really sure i get this.

Reply:
IArgumentOperation doesn't a property to the real argument expression. And I feel it should be its only child. (How one argument has multiple arguments?) But I am not 100% sure so I put the IsSingle() as a guard. Do I need to expain this in the code? Or if there is a better way to get the argument expression?

Reply:
Never mind. Just found that IArgumentOperation.Value is something that I want.

Reply:
:)